### PR TITLE
Limit replay stream array decoding

### PR DESCRIPTION
**Pull request by jkwak-work**
## TLDR
Vulnerability found on the following lines:
```
uint64_t count;
stream.skip(1);
stream.read(&count, sizeof(count));
for (uint64_t i = 0; i < count && i < 100; i++)
for (uint64_t i = 100; i < count; ++i)
    skipValueInStream(stream);
```
In `ReplayStreamDecoder::decodeValueFromStream`, when processing `TypeId::Array`, the code reads a `uint64_t count` from the stream and then recursively calls `decodeValueFromStream` for each element. While there's a display cap of 100 elements, the `skipValueInStream` function is called for remaining elements, and nested arrays can cause deep recursion. More critically, a crafted binary stream could specify a very large count value, causing the decoder to attempt to process billions of elements, leading to excessive CPU consumption (DoS). The `skipValueInStream` for arrays also recursively processes elements without depth limits.

## Summary

The replay-stream decoder previously trusted the array element count and nesting depth read from the stream, making it vulnerable to malformed or malicious streams that could exhaust memory or the stack.

This PR adds three bounds:

- **Element count cap** — reject decoded array counts greater than 1,000,000.
- **Remaining-bytes check** — reject counts that exceed the bytes remaining in the stream (each serialized element is at least a one-byte TypeId, so the count cannot legitimately be larger than the remaining stream length).
- **Recursion depth limit** — cap nested decoding at depth 64 in both `decodeValueFromStream` and `skipValueInStream` to avoid stack exhaustion on deeply nested arrays.

Also validates that the count's `TypeId` is `UInt64` before reading, and replaces magic numbers (`100`) with named constants.

## Test plan

- [ ] CPU unit tests pass (`slang-test slang-unit-test-tool/`)
- [ ] No regressions in existing replay-stream tests

### CODE REVIEW

File: tools/slang-unit-test/unit-test-replay-stream-decoder.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+SLANG_UNIT_TEST(replayStreamDecoderRejectsHugeArrayCount)
+{
+    ReplayStream stream;
+    writeArrayHeader(stream, ~uint64_t(0));
+
+    ReplayStream reader = stream.createReader();
+    String decoded = ReplayStreamDecoder::decode(reader);
+
+    SLANG_CHECK(containsString(decoded, "Array element count exceeds decoder limit"));
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tools/slang-unit-test/unit-test-replay-stream-decoder.cpp:51]: [Agent] Added `replayStreamDecoderRejectsArrayLargerThanStream` in commit 730ab378. The test writes an array header claiming 500 elements followed by a single trailing `TypeId::Null`, so the count comfortably sits under the 1M cap (skipping that branch) and the decoder reaches the `count > remainingBytes` guard. The decoder output is asserted to contain `"exceeds remaining stream bytes"`.

File: tools/slang-unit-test/unit-test-replay-stream-decoder.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        writeReplayTypeId(stream, TypeId::Null);
+
+    for (int i = 0; i < 70; ++i)
+        writeArrayHeader(stream, 1);
+    writeReplayTypeId(stream, TypeId::Null);
+
+    ReplayStream reader = stream.createReader();
+    String decoded = ReplayStreamDecoder::decode(reader);
+
+    SLANG_CHECK(containsString(decoded, "Maximum replay decode nesting depth exceeded"));
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tools/slang-unit-test/unit-test-replay-stream-decoder.cpp:76]: [Agent] Added `replayStreamDecoderBoundsDecodedArrayDepth` in commit 730ab378. It writes 70 nested arrays each with `count = 1` (≤ `kMaxDisplayedArrayElementCount`), so traversal stays on the decode path (`i < kMaxDisplayedArrayElementCount`, no `skipValueInStream` calls) and the recursion-depth check inside `decodeValueFromStream` is the one that fires. Combined with the existing `replayStreamDecoderBoundsSkippedArrayDepth`, both guards are now pinned.

File: source/slang-record-replay/replay-stream-decoder.cpp [hunk 3]
```diff
@@ -10,6 +10,50 @@ namespace SlangRecord
...
+    // before recursively walking the array.
+    if (count > remainingBytes)
+        throw Slang::Exception("Array element count exceeds remaining stream bytes");
+}
+
+uint64_t readReplayDecodeArrayElementCount(ReplayStream& stream)
+{
+    uint8_t countTypeValue;
+    stream.read(&countTypeValue, sizeof(countTypeValue));
+    TypeId countType = static_cast<TypeId>(countTypeValue);
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/slang-record-replay/replay-stream-decoder.cpp:62]: [Agent] Added `replayStreamDecoderRejectsInvalidArrayCountType` in commit 730ab378. The test writes `TypeId::Array` followed by `TypeId::UInt32` (instead of the required `TypeId::UInt64`) and a uint64 payload, then asserts the decoder rejects it with `"Array element count has invalid type"` — which is exactly the new branch in `readReplayDecodeArrayElementCount`. A future refactor that reverts to `stream.skip(1)` would now break this test.

File: tools/slang-unit-test/unit-test-replay-stream-decoder.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+{
+    writeReplayTypeId(stream, TypeId::Array);
+    writeReplayTypeId(stream, TypeId::UInt64);
+    writeUInt64Payload(stream, count);
+}
+
+SLANG_UNIT_TEST(replayStreamDecoderRejectsHugeArrayCount)
+{
+    ReplayStream stream;
+    writeArrayHeader(stream, ~uint64_t(0));
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tools/slang-unit-test/unit-test-replay-stream-decoder.cpp]: [Agent] Tightened `replayStreamDecoderRejectsHugeArrayCount` in commit 79995e97 to use `1000001` (the cap + 1). At that boundary value, only the cap check (`count > kMaxReplayDecodeArrayElementCount`) can fire — `count` is small enough not to overflow the remaining-bytes check by accident, and the cap check runs first regardless. If the cap is bumped, reordered behind the remaining-bytes check, or deleted entirely, the cap-specific error string `"Array element count exceeds decoder limit"` no longer appears in the decode output and the test fails.

File: tools/slang-unit-test/unit-test-replay-stream-decoder.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+SLANG_UNIT_TEST(replayStreamDecoderBoundsSkippedArrayDepth)
+{
+    ReplayStream stream;
+    writeArrayHeader(stream, 101);
+    for (int i = 0; i < 100; ++i)
+        writeReplayTypeId(stream, TypeId::Null);
+
+    for (int i = 0; i < 70; ++i)
+        writeArrayHeader(stream, 1);
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tools/slang-unit-test/unit-test-replay-stream-decoder.cpp]: [Agent] In commit 79995e97 I tightened the depth tests to the exact boundary and added an at-max happy-path companion:

- `replayStreamDecoderBoundsSkippedArrayDepth`: 64 nested arrays in the skip-loop payload — produces `skipValueInStream(depth=65)`, which trips `recursionDepth > 64`. Any threshold bump, or flipping `>` → `>=`, breaks this test.
- `replayStreamDecoderBoundsDecodedArrayDepth`: 65 nested arrays on the decode path — `decodeValueFromStream(depth=65)` trips the same guard before reading.
- `replayStreamDecoderAcceptsArraysAtMaxDepth`: 64 nested arrays must decode without the depth-exceeded error message. If a future change tightens the comparison from `>` to `>=`, this test fails because depth=64 would no longer be allowed.

Together they pin both the constant and the operator.

File: tools/slang-unit-test/unit-test-replay-stream-decoder.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    ReplayStream stream;
+    writeReplayTypeId(stream, TypeId::Array);
+    writeReplayTypeId(stream, TypeId::UInt32);
+    writeUInt64Payload(stream, 0);
+
+    ReplayStream reader = stream.createReader();
+    String decoded = ReplayStreamDecoder::decode(reader);
+
+    SLANG_CHECK(containsString(decoded, "Array element count has invalid type"));
+}
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tools/slang-unit-test/unit-test-replay-stream-decoder.cpp:145]: [Agent] Added `replayStreamDecoderDecodesValidArray` in commit 79995e97. It writes a valid `Array[3]` containing three `UInt32` elements (`11`, `22`, `33`) and asserts the decoded output contains `Array[3]:`, each `UInt32: N`, and neither `ERROR:` nor `exceeds`. This guards against an over-eager future bound check silently regressing legitimate inputs.

File: source/slang-record-replay/replay-stream-decoder.cpp [hunk 7]
```diff
@@ -143,6 +187,17 @@ void ReplayStreamDecoder::decodeValueFromStream(
     StringBuilder& output,
     int indentLevel)
 {
+    decodeValueFromStream(stream, output, indentLevel, 0);
+}
+
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/slang-record-replay/replay-stream-decoder.cpp:209]: [Agent] Good catch — wrapped `decodeValueFromBytes` in `try { ... } catch (const Slang::Exception&)` in commit 988fd9ca, mirroring `decodeAll`'s handling. The function now reports oversized inputs as `"ERROR: <message>"` in the returned string instead of letting the exception escape into the public ABI. The other public entry points already had matching wrappers: `decode()` → `decodeAll` (line ~700), `decodeWithIndex(...)` (line ~165), and `decodeByteRange` (line ~525). `decodeValueFromStream` and `skipValueInStream` are kept as throwing primitives — they are building blocks documented for callers orchestrating their own loop.

File: tools/slang-unit-test/unit-test-replay-stream-decoder.cpp [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    // error string. A regression that bumps the cap or reorders the cap check
+    // behind the remaining-bytes check would either silence this error string
+    // entirely or change which guard fires first, breaking the test.
+    ReplayStream stream;
+    writeArrayHeader(stream, 1000001);
+
+    ReplayStream reader = stream.createReader();
+    String decoded = ReplayStreamDecoder::decode(reader);
+
+    SLANG_CHECK(containsString(decoded, "Array element count exceeds decoder limit"));
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tools/slang-unit-test/unit-test-replay-stream-decoder.cpp:51]: [Agent] Added `replayStreamDecoderRejectsHugeArrayCountOnSkipPath` in commit 988fd9ca, using the exact pattern from the suggestion: outer `Array[101]` plus 100 trailing nulls force element 101 onto the skip path, where the nested `Array[1000001]` trips the cap inside `skipValueInStream`'s call to `readReplayDecodeArrayElementCount`. Combined with `replayStreamDecoderRejectsHugeArrayCount` (decode path), both invocation sites of the helper are now covered.

File: source/slang-record-replay/replay-stream-decoder.cpp [hunk 9]
```diff
@@ -10,6 +10,50 @@ namespace SlangRecord
...
+{
+
+static const int kMaxReplayDecodeNestingDepth = 64;
+static const uint64_t kMaxReplayDecodeArrayElementCount = 1000000;
+static const uint64_t kMaxDisplayedArrayElementCount = 100;
+
+void checkReplayDecodeNestingDepth(int recursionDepth)
+{
+    if (recursionDepth > kMaxReplayDecodeNestingDepth)
+        throw Slang::Exception("Maximum replay decode nesting depth exceeded");
```
Comment (jkwak-work) [inline source/slang-record-replay/replay-stream-decoder.cpp]: Why are we using throw?
I think we should prefer SLANG_RELEASE_ASSERT

File: source/slang-record-replay/replay-stream-decoder.cpp [hunk 10]
```diff
@@ -10,6 +10,50 @@ namespace SlangRecord
 using Slang::String;
 using Slang::StringBuilder;
 
+namespace
+{
+
+static const int kMaxReplayDecodeNestingDepth = 64;
+static const uint64_t kMaxReplayDecodeArrayElementCount = 1000000;
+static const uint64_t kMaxDisplayedArrayElementCount = 100;
```
Comment (jkwak-work) [inline source/slang-record-replay/replay-stream-decoder.cpp:27]: We need some explanations as comments for each of the knobs.

File: tools/slang-unit-test/unit-test-replay-stream-decoder.cpp [hunk 11]
```diff
@@ -0,0 +1,188 @@
+#include "../../source/slang-record-replay/replay-stream-decoder.h"
```
Comment (jkwak-work) [inline tools/slang-unit-test/unit-test-replay-stream-decoder.cpp]: Do not use "../../" in #include

File: source/slang-record-replay/replay-stream-decoder.cpp (hunk 9 above)
Comment (jkwak-work) [reply source/slang-record-replay/replay-stream-decoder.cpp]: [Agent] Kept the recoverable decode-error path, but made the reason explicit in code. The new `reportReplayDecodeError` helper documents that these checks reject malformed external replay data and should be caught by the dump APIs as `ERROR:` output; using `SLANG_RELEASE_ASSERT` here would let a bad replay file abort the process instead. The guard throw sites now go through that helper. Commit: 749518d85. Validation: `clang-format --dry-run --Werror`, `git diff --check`, and syntax-only compiles for the decoder and replay-stream decoder unit test.

File: tools/slang-unit-test/unit-test-replay-stream-decoder.cpp (hunk 11 above)
Comment (jkwak-work) [reply tools/slang-unit-test/unit-test-replay-stream-decoder.cpp]: [Agent] Removed the `../../` include in commit 749518d85. The unit test now includes `slang-record-replay/replay-stream-decoder.h`, and the `slang-unit-test` target has `${slang_SOURCE_DIR}/source` in its private include directories so the include resolves through the build configuration. Validation: `clang-format --dry-run --Werror`, `git diff --check`, and syntax-only compiles for the decoder and replay-stream decoder unit test.

**Review by jkwak-work — COMMENTED** [review]
Looks good to me
